python/lib/scheduler/dmod/scheduler/resources/resource_manager.py

```diff
@@ -170,126 +184,159 @@ def allocate_single_node(self, cpus: int, memory: int) -> List[Optional[Resource
+            with up to num_node-1 remainders to fill in.
+            Note that this is not the most generic "round robin" in which we allocate
             cpus one after the other across all available resources and don't try to balance.
             i.e. a request for 10 cpus with an available resource view of [4, 2, 4] would fail to allocate with this
-            algorithm, because it assumes an availablity of [4, 3, 3]
+            algorithm, because it assumes an availability of [4, 3, 3]
+
+            For a ``BUNDLED`` :class:`AllocationAssetGrouping`, a single allocation is returned per :class:`Resource`
+            node containing all the assets allocated from that node. For ``SILO``, several per-node allocations are
+            returned, each with a single CPU and a roughly even share of the total requested memory.
 
             Parameters
             ----------
                 cpus: Total number of CPUs requested
                 memory: Amount of memory required in bytes
+                asset_grouping: The way compute assets from a single node are grouped into allocations
 
             Returns
             -------
             [ResourceAlloction]
                 List of one or more ResourceAllocation if allocation successful, otherwise, [None]
         """
         #TODO consider scaling memory per cpu
-        #Find the number of cpus to allocate to each node
         self.validate_allocation_parameters(cpus, memory)
-        resources = list(self.get_useable_resources())
-        per_alloc_mem = int(memory / cpus)
-
-        num_node = len(resources)
-        if num_node == 0:
+        # Get all active and ready nodes, regardless of whether they are "allocateable" (i.e., not full)
+        resource_nodes = {r.resource_id: r for r in self.get_resources() if r.is_active() and r.is_ready()}
+
+        # Calculate in advance the exact amounts of CPUs and memory per node for the necessary balance
+        # This is slightly different from simply an even share due to discrete amounts and remainders
+        cpus_per_node, memory_per_node = dict(), dict()
+        num_nodes = len(resource_nodes)
+        cpu_share, mem_share = int(cpus / num_nodes), int(memory / num_nodes)
+        cpu_remainder, mem_remainder = cpus - cpu_share, memory - mem_share
+
+        for node_id, node in resource_nodes.items():
+            # We must plan to request at least the per-node amounts on each node
+            if node.cpu_count < cpu_share or node.memory < memory:
+                return [None]
+            cpus_per_node[node_id] = cpu_share
+            memory_per_node[node_id] = mem_share
+            # But we also need to try to pick up a share of the remainders if we can
+            if node.cpu_count > cpu_share and cpu_remainder > 0:
+                cpus_per_node[node_id] += 1
+                cpu_remainder -= 1
+            if node.memory > mem_share and mem_remainder > 0:
+                memory_per_node[node_id] += 1
```

Comment:
Let's table this.  I'm sure we can do better, but it's going to take a lot for that to be meaningful.  There are some problems with the way memory is being reserved and limited in the spawned Docker services right now, which is going to take some significant work to fix.  That likely is the time to address this.  

And regardless, I'm not entirely sure about the rounding:  do Docker reservations/limit round to the page?   I can't quickly find anything that says that they do.   That's not to say that we shouldn't expect to need these constraints; rather we may have to put them in at a different place (e.g., arg parsing), allowing us to assume only the easy-to-handle cases here.

Reply:
Yeah, that is fair. It will all boil down to cgroups at then end of the day. Here is a reference to the cgroup memory allocation documentation https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt